src/core/server/saved_objects/service/lib/saved_object_aggs_types/bucket_aggs/index.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import * as rt from 'io-ts';
+
+import { FieldBasicRT } from '../helpers';
+
+export const BucketAggsTypeRt = rt.partial({
+  filter: rt.type({
+    term: rt.record(rt.string, rt.string),
+  }),
+  histogram: rt.intersection([
+    FieldBasicRT,
+    rt.type({ interval: rt.number }),
```

Comment:
I get the 

> Anyway, the idea in this PR is that engineers who will need to add new aggregation types in src/core/server/saved_objects/service/lib/saved_object_aggs_types/*

However having the aggregations directly in `core/server` mean that:
-  3rd party plugins won't be able to add one and leverage this feature.
- `platform` is going to be the owner of these aggregations, instead of the providing teams.

I see that ATM all aggregations are used to generate a type

```ts
export type SavedObjectAggs = rt.TypeOf<typeof SavedObjectAggsRt>;
```

However my main question is: to which extend is this `SavedObjectAggs` useful / to which extend could we refactor this to allow plugins to register aggregation using `core` SO APIs instead of having to add them in `core` code base directly?

Reply:
I think that's not a bad idea but we still want to check/control the type of the aggregations because we do not want to add script attributes and be able to access another saved objects outside of the aggregations.
I do agree that it will be nice at one point to have other plugins to access these types, that I was my thought too but I was thinking we do not need to resolve everything at ones.

Reply:
> I mean SavedObjectAggs is our guard here.

Do you mean `SavedObjectAggs` or `SavedObjectAggsRt`? From what I understand:

`SavedObjectAggsRt` is necessary, but it could be generated the same way if the aggregations/schemas were registered from plugins during `setup` phase, while having the individual `aggs` and associated TS types kept in the registrant plugins code. They could even export these 'generated' TS type if another dependant plugin would need to use `agg` registered from another plugin at some point.

That would just need to have methods such as

```ts
export const validateSavedObjectTypeAggs = (aggObjects: SavedObjectAggs) => {
  pipe(
    excess(SavedObjectAggsRt).decode(aggObjects),
    fold(throwErrors(SavedObjectsErrorHelpers.createBadRequestError), identity)
  );
};
```

to have the `SavedObjectAggsRt` obj passed as a parameter instead of using direct static reference.

`SavedObjectAggs` OTOH seems only really used in utility methods such as `validateGetSavedObjectAggs` and could probable be changed to `Record<string, probablyAny>`

AFAIK it's already typed as `any` in this signature

```ts
// src/core/server/saved_objects/service/lib/aggs_utils.ts L35
const validateGetAggFieldValue = (
  allowedTypes: string[],
  aggs: any,
  indexMapping: IndexMapping
)
```

Also it's just parsed as `any` in

```ts
validateGetSavedObjectAggs(allowedTypes, JSON.parse(aggs), this._mappings);
```

Reply:
There isn't a safe way to blacklist `script` keys since some aggregations accept scripts in a key like `init_script` so we need a whitelist approach. This is unfortunate since we have to duplicate the entire aggregation framework to construct this whitelist.

To add a bit of context, Xavier intends to add more aggregations than what's in this PR if we agree on the approach. I think our goal should be to cover the entire aggregation API, however, depending on the amount of work this might not be feasible as a first step. We should definitely debate this, releasing something useful that might cover 80% of the use cases vs putting in the extra effort to getting to 100%. I don't have a good feeling for the amount of work.

But if we don't go to 100% it's relatively easy for other teams to make these contributions to Core so that it doesn't have to fall entirely on Platform. The Platform team will have to review all these PR's to make sure they're secure.